prometheus/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,135 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	"context"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"knative.dev/eventing/pkg/adapter"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/source"
+
+	"knative.dev/eventing-contrib/prometheus/pkg/apis/sources/v1alpha1"
+)
+
+const (
+	apiChunkOfURL = `/api/v1/query?query=`
+)
+
+type envConfig struct {
+	adapter.EnvConfig
+
+	EventSource string `envconfig:"EVENT_SOURCE" required:"true"`
+	ServerURL   string `envconfig:"PROMETHEUS_SERVER_URL" required:"true"`
+	PromQL      string `envconfig:"PROMETHEUS_PROM_QL" required:"true"`
+}
+
+type prometheusAdapter struct {
+	source    string
+	ce        cloudevents.Client
+	reporter  source.StatsReporter
+	namespace string
+	logger    *zap.SugaredLogger
+	serverURL string
+	promQL    string
+}
+
+func NewEnvConfig() adapter.EnvConfigAccessor {
+	return &envConfig{}
+}
+
+// NewAdapter creates an adapter to convert PromQL replies to CloudEvents
+func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client, reporter source.StatsReporter) adapter.Adapter {
+	logger := logging.FromContext(ctx)
+	env := processed.(*envConfig)
+
+	a := &prometheusAdapter{
+		source:    env.EventSource,
+		ce:        ceClient,
+		reporter:  reporter,
+		logger:    logger,
+		namespace: env.Namespace,
+		serverURL: env.ServerURL,
+		promQL:    env.PromQL,
+	}
+
+	return a
+}
+
+func (a *prometheusAdapter) Start(stopCh <-chan struct{}) error {
+	return wait.PollUntil(5*time.Second, a.send, stopCh)
+}
+
+func (a *prometheusAdapter) send() (done bool, err error) {
+	var completeURL = a.serverURL + apiChunkOfURL + a.promQL
+
+	req, err := http.NewRequest(`GET`, completeURL, nil)
+	if err != nil {
+		return false, err
+	}
+
+	a.logger.Info(req)
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	reply, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, err
+	}
+
+	if len(reply) > 0 {
+		event, err := a.makeEvent(reply)
+		if err != nil {
+			return false, err
+		}
+
+		if _, _, err := a.ce.Send(context.Background(), *event); err != nil {
+			a.logger.Info("event delivery failed", zap.Error(err))
+			return false, err
+		}
+	}
+
+	return false, nil
+}
+
+func (a *prometheusAdapter) makeEvent(payload interface{}) (*cloudevents.Event, error) {
+	event := cloudevents.NewEvent()
+	event.SetSource(a.source)
+	event.SetID(`prometheus_source_adapter_id`)
```

Comment:
Source + ID must be unique.

What about documenting it by adding the CloudEvents adapter spec for Prometheus? (see https://github.com/cloudevents/spec/blob/master/adapters/github.md for an example)

(I'm doing the same for CouchDB)